docs/resources/extract-ai.sh

```diff
@@ -6,26 +6,88 @@ BINARY_FOLDER=/var/mnt
 
 pushd $FOLDER
 
-total_copies=$(sort -u $BINARY_FOLDER/$OCP_RELEASE_LIST | wc -l)  # Required to keep track of the pull task vs total
-current_copy=1
+load_images() {
+
+declare -A pids # Hash that include the images pulled along with their pids to be monitored by wait command
+local max_bg=10 # Max number of simultaneous skopeo copies to container storage
+local total_copies=$(sort -u $BINARY_FOLDER/$OCP_RELEASE_LIST | wc -l)  # Required to keep track of the pull task vs total
+local current_copy=1
+
+#remove duplicates
+sort -u -o $OCP_RELEASE_LIST $OCP_RELEASE_LIST
 
 while read -r line;
 do
   uri=$(echo "$line" | awk '{print$1}')
-  #tar=$(echo "$line" | awk '{print$2}')
   podman image exists $uri
   if [[ $? -eq 0 ]]; then
-      echo "Skipping existing image $tar"
-      echo "Copying ${uri} [${current_copy}/${total_copies}]"
+      echo "[INFO] Skipping existing image $tar"
+      echo "[INFO] Copying ${uri} [${current_copy}/${total_copies}]"
       current_copy=$((current_copy + 1))
       continue
   fi
   tar=$(echo "$uri" |  rev | cut -d "/" -f1 | rev | tr ":" "_")
-  tar zxvf ${tar}.tgz
+  tar --use-compress-program=pigz -xf ${tar}.tgz
   if [ $? -eq 0 ]; then rm -f ${tar}.gz; fi
-  echo "Copying ${uri} [${current_copy}/${total_copies}]"
-  skopeo copy dir://$(pwd)/${tar} containers-storage:${uri}
-  if [ $? -eq 0 ]; then rm -rf ${tar}; current_copy=$((current_copy + 1)); fi
+  echo "[INFO] Copying ${uri} [${current_copy}/${total_copies}]"
+  skopeo copy dir://$(pwd)/${tar} containers-storage:${uri} -q &
+
+  pids[${uri}]=$! # Keeping track of the PID and container image in case the pull fails
+  max_bg=$((max_bg - 1)) # Batch size adapted 
+  current_copy=$((current_copy + 1)) 
+  if [[ $max_bg == 0 ]] || [[ $current_copy > $total_copies ]] # If the batch is done, then monitor the status of all pulls before moving to the next batch. If the last images are being pulled wait too.
+  then
+    for pid in ${!pids[@]}; do
+      wait ${pids[$pid]} # The way wait monitor for each background task (PID). If any error then copy the image in the failed array so it can be retried later
+      if [[ $? != 0 ]]; then
+        echo "[ERROR] Pull failed for container image: ${pid} . Retrying later... "
+        failed_copies+=(${pid}) # Failed, then add the image to be retrieved later
+      else
+        echo "[INFO] Removing folder for ${pid}"
+        targz=$(echo "$pid" |  rev | cut -d "/" -f1 | rev | tr ":" "_")
```

Comment:
I found this very confusing, thinking $pid was a PID, and couldn't figure out what it was trying to do. Can you rename this iterator var to something like $img instead?

Also, I think this command could be more efficient and a little clearer written as something like:
`targz=$(basename ${img/:/_})`

Each pipe in the current command creates a new subprocess. Along with the similar command at line 29, that's 8 subprocesses created for each image that could be avoided. Without the operators, I think we have a couple hundred images, so that's 1600 subprocesses we could avoid creating